File: python/src/browser_commander/downloads/watcher.py
"""A download source for engines with no download events (issue #88).

Selenium's CDP bridge is request/response only: it can tell Chromium where to
put downloads, but it has no event stream to hear about them on. Watching the
staging directory is what gives Selenium the same managed lifecycle - including
downloads a person started by hand, because ``Browser.setDownloadBehavior`` is
browser-wide.
"""

from __future__ import annotations

import asyncio
import contextlib
from pathlib import Path
from typing import Any, Callable

from browser_commander.downloads.destination import ARTIFACT_DIRECTORY_MODE
from browser_commander.downloads.sources import STAGING_DIRECTORY, SourceHandle
from browser_commander.downloads.store import DownloadSource

#: How often the staging directory is listed.
DEFAULT_POLL_INTERVAL = 0.25

#: Suffixes Chromium uses while a file is still being written.
IN_PROGRESS_SUFFIXES = (".crdownload", ".tmp", ".partial")


def _set_download_behavior(driver: Any, staging_directory: str) -> None:
    """Point a Selenium-driven Chromium at the staging directory.

    Args:
        driver: Selenium WebDriver
        staging_directory: Where Chromium should write downloads
    """
    driver.execute_cdp_cmd(
        "Browser.setDownloadBehavior",
        {"behavior": "allow", "downloadPath": staging_directory},
    )


class DirectoryWatcher:
    """Report files that appear in a staging directory and stop growing.

    A file is only reported once its size has stopped changing: Chromium
    renames ``report.pdf.crdownload`` to ``report.pdf`` when it is done, but a
    file that merely exists is not evidence that its last byte was written.
    """

    def __init__(self, *, directory: str, sink: Any) -> None:
        """Start watching a directory.

        Args:
            directory: Staging directory to watch
            sink: Object with ``started``/``finished``/``failed``/``track``
        """
        self._directory = directory
        self._sink = sink
        self._sizes: dict[str, int] = {}
        self._reported: set[str] = set()

        # Files that were already there belong to an earlier session; claiming
        # them would report downloads this manager never saw.
        for name in self._candidates():
            self._reported.add(name)

    def _candidates(self) -> list[str]:
        """List the files that could be finished downloads.

        Returns:
            File names, in a stable order
        """
        try:
            entries = sorted(Path(self._directory).iterdir())
        except OSError:
            return []

        return [
            entry.name
            for entry in entries
            if entry.is_file()
            and not entry.name.startswith(".")
            and not entry.name.endswith(IN_PROGRESS_SUFFIXES)
        ]

    def poll_once(self) -> list[str]:
        """Report every download that has finished since the last poll.

        Returns:
            Names reported by this poll
        """
        reported = []
        for name in self._candidates():
            if name in self._reported:
                continue

            path = Path(self._directory, name)
            try:
                size = path.stat().st_size
            except OSError:
                continue

            # The first sighting only records the size; a file is claimed on
            # the poll that finds it unchanged.
            if self._sizes.get(name) != size:
                self._sizes[name] = size
                continue

            self._reported.add(name)
            self._sizes.pop(name, None)
            record = self._sink.started(
                engine_handle=str(path),
                url=path.as_uri(),
                suggested_filename=name,
            )
            self._sink.track(
                self._sink.finished(
                    record, DownloadSource(path=str(path), remove_source=True)
                )
            )
            reported.append(name)
        return reported


async def attach_filesystem_watcher(
    *,
    root: str,
    sink: Any,
    driver: Any = None,
    set_behavior: Callable[[str], None] | None = None,
    interval: float = DEFAULT_POLL_INTERVAL,
) -> SourceHandle:
    """Send downloads to a staging directory and watch it for finished files.

    Args:
        root: Managed download directory
        sink: Object with ``started``/``finished``/``failed``/``track``
        driver: Selenium WebDriver used to configure download behavior
        set_behavior: Override for the CDP call, used by tests
        interval: Seconds between directory listings

    Returns:
        A handle that stops the watcher
    """
    staging_directory = str(Path(root, STAGING_DIRECTORY))
    Path(staging_directory).mkdir(
        parents=True, exist_ok=True, mode=ARTIFACT_DIRECTORY_MODE
    )

    configure = set_behavior or (
        (lambda directory: _set_download_behavior(driver, directory))
        if driver is not None
        else None
    )
    if configure is not None:
        configure(staging_directory)

    watcher = DirectoryWatcher(directory=staging_directory, sink=sink)

    async def loop() -> None:
        while True:
            await asyncio.sleep(interval)
            watcher.poll_once()

    task = asyncio.ensure_future(loop())

    async def detach() -> None:
        task.cancel()
        with contextlib.suppress(asyncio.CancelledError):
            await task
        # One last look: a download that finished between the final poll and
        # this call is still the caller's file.
        watcher.poll_once()

    return SourceHandle(
        detach=detach, staging_directory=staging_directory, watcher=watcher
    )

File: rust/src/downloads/watcher.rs
//! A download source for engines with no download events (issue #88).
//!
//! Rust's CDP transport is request/response only: it can tell Chromium where to
//! put downloads, but it has no event stream to hear about them on, and
//! `fantoccini` has neither. Watching the staging directory is what gives both
//! the same managed lifecycle — including downloads a person started by hand,
//! because `Browser.setDownloadBehavior` is browser-wide.
//!
//! # Example
//!
//! ```rust
//! use browser_commander::downloads::DirectoryWatcher;
//!
//! let staging = std::env::temp_dir().join("bc-watcher-doc");
//! std::fs::create_dir_all(&staging)?;
//! let mut watcher = DirectoryWatcher::new(&staging);
//!
//! std::fs::write(staging.join("report.pdf"), b"%PDF-1.7")?;
//! // The first sighting only measures the file; a download is claimed once
//! // its size has stopped changing.
//! assert!(watcher.poll_once().is_empty());
//! assert_eq!(watcher.poll_once().len(), 1);
//! # std::fs::remove_dir_all(&staging)?;
//! # Ok::<(), std::io::Error>(())
//! ```

use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::time::Duration;

use crate::downloads::sources::{
    prepare_staging_directory, DownloadSink, DownloadStart, SourceHandle,
};
use crate::downloads::store::DownloadSource;
use crate::downloads::DownloadError;

/// How often the staging directory is listed.
pub const DEFAULT_POLL_INTERVAL: Duration = Duration::from_millis(250);

/// Suffixes Chromium uses while a file is still being written.
pub const IN_PROGRESS_SUFFIXES: [&str; 3] = ["crdownload", "tmp", "partial"];

/// A finished download found in the staging directory.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StagedDownload {
    /// Where the browser left the bytes.
    pub path: PathBuf,
    /// The name the browser gave the file, which is the name the page suggested.
    pub name: String,
}

/// Report files that appear in a staging directory and stop growing.
///
/// A file is only reported once its size has stopped changing: Chromium renames
/// `report.pdf.crdownload` to `report.pdf` when it is done, but a file that
/// merely exists is not evidence that its last byte was written.
#[derive(Debug)]
pub struct DirectoryWatcher {
    directory: PathBuf,
    sizes: HashMap<String, u64>,
    reported: HashSet<String>,
}

impl DirectoryWatcher {
    /// Start watching a directory.
    ///
    /// # Arguments
    ///
    /// * `directory` - Staging directory to watch
    ///
    /// # Returns
    ///
    /// A watcher that ignores everything already in the directory.
    pub fn new(directory: impl Into<PathBuf>) -> Self {
        let directory = directory.into();
        // Files that were already there belong to an earlier session; claiming
        // them would report downloads this manager never saw.
        let reported = Self::candidates(&directory).into_iter().collect();
        Self {
            directory,
            sizes: HashMap::new(),
            reported,
        }
    }

    /// List the files that could be finished downloads.
    fn candidates(directory: &Path) -> Vec<String> {
        let Ok(entries) = std::fs::read_dir(directory) else {
            return Vec::new();
        };

        let mut names: Vec<String> = entries
            .filter_map(Result::ok)
            .filter(|entry| entry.path().is_file())
            .filter_map(|entry| entry.file_name().into_string().ok())
            .filter(|name| !name.starts_with('.'))
            .filter(|name| {
                !IN_PROGRESS_SUFFIXES
                    .iter()
                    .any(|suffix| name.ends_with(&format!(".{suffix}")))
            })
            .collect();
        names.sort();
        names
    }

    /// Report every download that has finished since the last poll.
    ///
    /// # Returns
    ///
    /// The downloads claimed by this poll, in a stable order.
    pub fn poll_once(&mut self) -> Vec<StagedDownload> {
        let mut claimed = Vec::new();

        for name in Self::candidates(&self.directory) {
            if self.reported.contains(&name) {
                continue;
            }

            let path = self.directory.join(&name);
            let Ok(size) = std::fs::metadata(&path).map(|metadata| metadata.len()) else {
                continue;
            };

            // The first sighting only records the size; a file is claimed on
            // the poll that finds it unchanged.
            if self.sizes.get(&name) != Some(&size) {
                self.sizes.insert(name, size);
                continue;
            }

            self.sizes.remove(&name);
            self.reported.insert(name.clone());
            claimed.push(StagedDownload { path, name });
        }

        claimed
    }
}

/// Watch a staging directory and hand every finished download to a sink.
///
/// # Arguments
///
/// * `root` - Managed download directory
/// * `sink` - Where downloads are reported, usually the manager
/// * `interval` - How long to wait between directory listings
///
/// # Returns
///
/// A handle that stops the watcher.
///
/// # Errors
///
/// Returns [`DownloadError::DirectoryNotCreated`] when the staging directory
/// cannot be created.
pub fn attach_filesystem_watcher(
    root: &Path,
    sink: Arc<dyn DownloadSink>,
    interval: Duration,
) -> Result<SourceHandle, DownloadError> {
    let staging_directory = prepare_staging_directory(root)?;
    let mut watcher = DirectoryWatcher::new(&staging_directory);
    let (stop, mut stopped) = tokio::sync::oneshot::channel::<()>();

    let task = tokio::spawn({
        let sink = Arc::clone(&sink);
        async move {
            loop {
                tokio::select! {
                    _ = &mut stopped => break,
                    _ = tokio::time::sleep(interval) => {}
                }
                report(&mut watcher, &sink).await;
            }
            // One last look: a download that finished between the final poll
            // and this call is still the caller's file.
            report(&mut watcher, &sink).await;
        }
    });

    Ok(SourceHandle {
        staging_directory,
        detach: Box::new(move || {
            Box::pin(async move {
                let _ = stop.send(());
                let _ = task.await;
            })
        }),
    })
}

/// Hand one poll's worth of finished downloads to the sink.
async fn report(watcher: &mut DirectoryWatcher, sink: &Arc<dyn DownloadSink>) {
    for staged in watcher.poll_once() {
        let id = sink.started(DownloadStart {
            engine_handle: staged.path.to_string_lossy().into_owned(),
            url: Some(format!("file://{}", staged.path.to_string_lossy())),
            suggested_filename: Some(staged.name),
            mime_type: None,
        });
        // Awaited rather than spawned: the watcher's loop is the only thing
        // holding this work, so returning before the bytes are placed would
        // let `dispose()` finish with a `.partial` file as the only trace.
        sink.finished(id, DownloadSource::staged(staged.path)).await;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::downloads::test_support::{RecordingSink, TempDir};

    /// Write `contents` into the staging directory.
    fn write(directory: &Path, name: &str, contents: &[u8]) {
        std::fs::write(directory.join(name), contents).unwrap();
    }

    #[test]
    fn claims_a_file_only_once_it_has_stopped_growing() {
        let temp = TempDir::new("bc-watch-growing");
        let mut watcher = DirectoryWatcher::new(temp.path());

        write(temp.path(), "report.pdf", b"half");
        assert!(watcher.poll_once().is_empty(), "claimed a growing file");

        write(temp.path(), "report.pdf", b"half and the rest");
        assert!(watcher.poll_once().is_empty(), "claimed a growing file");

        let claimed = watcher.poll_once();
        assert_eq!(claimed.len(), 1);
        assert_eq!(claimed[0].name, "report.pdf");
        assert_eq!(claimed[0].path, temp.path().join("report.pdf"));
    }

    #[test]
    fn claims_a_download_exactly_once() {
        let temp = TempDir::new("bc-watch-once");
        let mut watcher = DirectoryWatcher::new(temp.path());

        write(temp.path(), "report.pdf", b"body");
        watcher.poll_once();
        assert_eq!(watcher.poll_once().len(), 1);
        assert!(
            watcher.poll_once().is_empty(),
            "claimed the same file twice"
        );
    }

    #[test]
    fn ignores_files_chromium_is_still_writing() {
        let temp = TempDir::new("bc-watch-inprogress");
        let mut watcher = DirectoryWatcher::new(temp.path());

        write(temp.path(), "report.pdf.crdownload", b"body");
        write(temp.path(), "notes.txt.tmp", b"body");
        write(temp.path(), ".hidden", b"body");
        watcher.poll_once();

        assert!(watcher.poll_once().is_empty());
    }

    #[test]
    fn ignores_files_that_were_there_before_the_session() {
        let temp = TempDir::new("bc-watch-existing");
        write(temp.path(), "from-yesterday.pdf", b"body");

        let mut watcher = DirectoryWatcher::new(temp.path());
        watcher.poll_once();

        // Reporting a file this manager never saw arrive would claim somebody
        // else's download as this session's.
        assert!(watcher.poll_once().is_empty());
    }

    #[test]
    fn survives_a_directory_that_is_not_there() {
        let temp = TempDir::new("bc-watch-missing");
        let mut watcher = DirectoryWatcher::new(temp.path().join("never-created"));

        assert!(watcher.poll_once().is_empty());
    }

    #[tokio::test]
    async fn saves_what_the_browser_left_in_the_staging_directory() {
        let temp = TempDir::new("bc-watch-attach");
        let root = temp.path().join("downloads");
        let sink = Arc::new(RecordingSink::default());

        let handle = attach_filesystem_watcher(
            &root,
            Arc::clone(&sink) as Arc<dyn DownloadSink>,
            Duration::from_millis(10),
        )
        .unwrap();
        write(&handle.staging_directory, "report.pdf", b"%PDF-1.7");
        tokio::time::sleep(Duration::from_millis(60)).await;
        (handle.detach)().await;

        let seen = sink.started_names();
        assert_eq!(seen, vec!["report.pdf".to_string()]);
        assert_eq!(sink.finished_count(), 1);
    }

    #[tokio::test]
    async fn takes_one_last_look_when_it_is_detached() {
        let temp = TempDir::new("bc-watch-last-look");
        let root = temp.path().join("downloads");
        let sink = Arc::new(RecordingSink::default());

        let handle = attach_filesystem_watcher(
            &root,
            Arc::clone(&sink) as Arc<dyn DownloadSink>,
            Duration::from_millis(50),
        )
        .unwrap();
        // One poll measures the file, and the next one would claim it; the
        // detach happens first, so only the final look can report it.
        write(&handle.staging_directory, "late.pdf", b"body");
        tokio::time::sleep(Duration::from_millis(80)).await;
        (handle.detach)().await;

        assert_eq!(sink.started_names(), vec!["late.pdf".to_string()]);
    }
}
